### DISCUSSION TITLE
Using ruff as a Python library

### DISCUSSION THREAD

**Discussion by jgarte** [astral-sh/ruff#8539] [Q&A]
Hi,

Is there currently a supported API for using ruff as a Python library? I'm assuming yes, because of the affordances of [pyo3](https://pyo3.rs/v0.20.0/getting_started#running-code)...

Maybe the better question is, should we be documenting how to use ruff as a Python library?

My use case is the following scenario:

pydocstyle had a way to import `pydocstyle` as a library and check code. So, I could write a pydocstyle script with custom logic to check a module or collection of files. See below for an example of importing pydocstyle API functions in a script:

https://github.com/databand-ai/dbnd/blob/63fb610d1b43c5e512ba66eba238dbbd210acb1f/docs/check_valid_docstring.py#L86

or here

https://github.com/dfrtz/textology/blob/878e6d8297a43082181421b88c09933f79af70c5/utils/pydocstyle_patched.py#L15

WDYT

**Comment by zanieb** [comment]
We actually don't support a Python API nor do we support custom extensions.

We intend to in the future, but there is a lot of design work that needs to happen first.

Some more context in https://github.com/astral-sh/ruff/issues/1992

**Comment by akx** [comment]
If it helps, you can pretty easily use `subprocess` for the time being, see https://github.com/astral-sh/ruff/issues/8401#issuecomment-1788806462 (and e.g. https://github.com/astral-sh/ruff/issues/8673).
